sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+#define KL_LOG KL_DEV
+
+/* For detailed information on the controller, please refer to:
+ * https://cs140e.sergio.bz/docs/BCM2837-ARM-Peripherals.pdf
+ */
+
+#include <sys/mimiker.h>
+#include <sys/rman.h>
+#include <sys/devclass.h>
+#include <sys/klog.h>
+#include <sys/types.h>
+#include <dev/bcm2835_gpio.h>
+#include <sys/device.h>
+#include <sys/bus.h>
+#include <sys/time.h>
+#include <sys/libkern.h>
+#include <sys/callout.h>
+#include <sys/condvar.h>
+#include <sys/bus.h>
+#include <dev/emmc.h>
+#include <sys/errno.h>
+#include <sys/bitops.h>
+
+/* Registers */
+#define BCMEMMC_ARG2 0x0000
+#define BCMEMMC_BLKSIZECNT 0x0004
+#define BCMEMMC_ARG1 0x0008
+#define BCMEMMC_CMDTM 0x000C
+#define BCMEMMC_RESP0 0x0010
+#define BCMEMMC_RESP1 0x0014
+#define BCMEMMC_RESP2 0x0018
+#define BCMEMMC_RESP3 0x001C
+#define BCMEMMC_DATA 0x0020
+#define BCMEMMC_STATUS 0x0024
+#define BCMEMMC_CONTROL0 0x0028
+#define BCMEMMC_CONTROL1 0x002C
+#define BCMEMMC_INTERRUPT 0x0030
+#define BCMEMMC_INT_MASK 0x0034
+#define BCMEMMC_INT_EN 0x0038
+#define BCMEMMC_CONTROL2 0x003C
+#define BCMEMMC_SLOTISR_VER 0x00FC
+
+/* STATUS register fields */
+#define SR_DAT_INHIBIT 0x00000002
+#define SR_CMD_INHIBIT 0x00000001
+
+/* INTERRUPT register fields */
+#define INT_DATA_TIMEOUT 0x00100000
+#define INT_CMD_TIMEOUT 0x00010000
+#define INT_READ_RDY 0x00000020
+#define INT_WRITE_RDY 0x00000010
+#define INT_DATA_DONE 0x00000002
+#define INT_CMD_DONE 0x00000001
+
+#define INT_ERROR_MASK 0x017E8000
+
+/* CONTROL register fields */
+#define C0_HCTL_DWITDH 0x00000002
+#define C0_HCTL_8BIT 0x00000020
+
+#define C1_SRST_HC 0x01000000
+#define C1_TOUNIT_MAX 0x000e0000
+#define C1_CLK_EN 0x00000004
+#define C1_CLK_STABLE 0x00000002
+#define C1_CLK_INTLEN 0x00000001
+
+/* SLOTISR_VER values */
+#define HOST_SPEC_NUM 0x00ff0000
+#define HOST_SPEC_NUM_SHIFT 16
+#define HOST_SPEC_V3 2
+#define HOST_SPEC_V2 1
+#define HOST_SPEC_V1 0
+
+static driver_t bcmemmc_driver;
```

Comment:
That's a weird one (see other comments regarding its usage).

Reply:
I'm not sure what you mean.

Reply:
Ok, I removed it after getting rid of some asserts.